### PR TITLE
Modernize your pipeline to be arm ready with multi arch container builds

### PR DESCRIPTION
**Pull request by lward27**
Ready for Reviews!! Modernize your pipeline to be arm ready with multi-arch container builds. Co-authored by John Blat and Lucas Ward. Trying to publish this one tomorrow by lunch, please help!

### CODE REVIEW

File: posts/Modernize-your-pipeline-to-be-ARM-ready-with-multi-arch-container-builds.md [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+
+# Where this work happens
+
+## Either way, it's turtles all the way down<sup>1</sup>
+
+It may be confusing at first to understand exactly what it means to convert your existing workloads to be able to run on multiple architectures. Let us look at a gross oversimplification.... All code inevitably ends up being ran as machine code that uses a given CPU's instruction set. In other words, it's all 1s and 0s eventually. For x86_64 and ARM, look at the machine code that is running for each architecture and it will be different.
+
+The cool part about this whole thing is, changes in your application code may not even be required. Often times it is as easy as using a different compiler flag, running interpreted code with an interpreter compiled for a different architecture, or instructing `docker buildx` to build an image for a different architecutre. That's right, these changes are made during a phase of the software development lifecycle called the release phase, specifically the "build process". For this article, we will be focused on a build process that takes place in a build pipeline, namely, gitlab pipelines.
```
Comment (jscharf-ippon) [inline posts/Modernize-your-pipeline-to-be-ARM-ready-with-multi-arch-container-builds.md:39]: This sentence feels too wordy

That's right, these changes are made during a phase of the software development lifecycle called the release phase, specifically the "build process".

File: posts/Modernize-your-pipeline-to-be-ARM-ready-with-multi-arch-container-builds.md [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+COPY ./print-platform-info.py /print-platform-info
+CMD ["python3", "/print-platform-info"]
+```
+Now this is typically where you would scamper off to the command line and run `docker build -t my_image .`, but hold off for just a minute! We have to set docker to use a different *build instance*. In the command line inside the project directory where you have your `Dockerfile` run:
+```bash
+docker buildx create --use
+```
+You may need to add some additional arguments to authenticate to your docker registy in the above command. Check the [docs](https://docs.docker.com/engine/reference/commandline/buildx_use/) for any information or google any errors you may run into. Now it is time to run the actual build command. Back at the command prompt, run:
+```bash
+`docker buildx build --platform linux/arm64,linux/amd64 -t <registry-url>/<image-name>:<image-tag> . --push`
```
Comment (jscharf-ippon) [inline posts/Modernize-your-pipeline-to-be-ARM-ready-with-multi-arch-container-builds.md:112]: too many back ticks here since you are already in a bash block you dont need extra

File: posts/Modernize-your-pipeline-to-be-ARM-ready-with-multi-arch-container-builds.md [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+```
+Let's take a look at the different pieces of this command...
+* `docker buildx build` - This simply tells docker to use the buildx build instance that we set before.
```
Comment (jscharf-ippon) [inline posts/Modernize-your-pipeline-to-be-ARM-ready-with-multi-arch-container-builds.md:115]: This tells docker to start a command using buildkit. There is a separate command `--builder` that lets you override the builder

File: posts/Modernize-your-pipeline-to-be-ARM-ready-with-multi-arch-container-builds.md [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+* `--platform linux/arm64,linux/amd64` - tells buildx to produce an image for arm64 and amd64. The builder instance actually pulls down the native architecture image to build on.
+* `-t <registry-url>/<iamge-name>:<image-tag>'` - This is where you specify your registry url and give your image a name and a tag. If you don't specify a registry, one of your images may end up living in the docker cache.
+* `. --push` - tells docker to use the local directory Dockerfile and then push the built images to our registry.
```
Comment (jscharf-ippon) [inline posts/Modernize-your-pipeline-to-be-ARM-ready-with-multi-arch-container-builds.md:118]: . has to do with the directory that is tar and sent to the docker daemon, which in turn if there is a "Dockerfile" in that directory it will get picked up automatically.

**Review by jscharf-ippon — CHANGES_REQUESTED** [review]

File: posts/Modernize-your-pipeline-to-be-ARM-ready-with-multi-arch-container-builds.md (hunk 1 above)
Comment (jscharf-ippon) [reply posts/Modernize-your-pipeline-to-be-ARM-ready-with-multi-arch-container-builds.md:39]: Is the build phase part of the release phase?

File: posts/Modernize-your-pipeline-to-be-ARM-ready-with-multi-arch-container-builds.md [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+  - aarch64
+  - amd64
+  - Ampere Altra
+  - Pipeline
+  - GitLab
+date: 2023-03-17T00:00:00.000Z
+title: "Modernize your pipeline to be ARM-ready with multi-arch container builds"
+---
+
+ARM computing has been historically reserved for embedded and small portable devices, however, in recent years there has been a growing number of ARM products in cloud and laptop products. This includes AWS Graviton Compute Instances, Raspberry Pi's Cortex-A72 chip, Googles Tau T2A compute instances, Azures Ampre Alta machines, and the pivot of Apple's ecosystem to ARM based processors. I know what you are thinking, Graviton, Tau, and Ampre Alta are really cool names, I agree! Given this new wave of ARM products, it is important to support the ARM architecture with multi-arch container images. Converting your pipeline to build multi-arch images is a relatively straightforward and quick process that pays off in a larger user base and higher cost savings. At Ippon Technologies, we utilize docker's buildx plugin to achieve building multi-arch images. Below is a short getting started guide, as well as some discussion as why, as a developer, it is in your best interest to build multi-arch images.
```
Comment (fhatfield-ippon) [inline posts/Modernize-your-pipeline-to-be-ARM-ready-with-multi-arch-container-builds.md:19]: Also, "I know what you are thinking: Graviton, Tau, and Ampre Alta are really cool names. I agree!" un-run-ons this sentence.

File: posts/Modernize-your-pipeline-to-be-ARM-ready-with-multi-arch-container-builds.md [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+
+The cool part about this whole thing is, changes in your application code may not even be required. Often times it is as easy as using a different compiler flag, running interpreted code with an interpreter compiled for a different architecture, or instructing `docker buildx` to build an image for a different architecutre. That's right, these changes are made during a phase of the software development lifecycle called the release phase, specifically the "build process". For this article, we will be focused on a build process that takes place in a build pipeline, namely, gitlab pipelines. 
+
+# What are the steps
+
+For some context, we will frame the following example with a Kubernetes cluster running on Amazon Web Service's EKS. For brevities sake, I am going to assume that you already have a Kubernetes cluster running and some facility operating it. If you don't, here is a very brief version of how to get up and running. Please note, this is not the only application for running multi-arch images, but definitely a relevant one if you live and breath DevOps work as your day to day.
```
Comment (fhatfield-ippon) [inline posts/Modernize-your-pipeline-to-be-ARM-ready-with-multi-arch-container-builds.md:43]: "brevity's", "brevities" is plural.

File: posts/Modernize-your-pipeline-to-be-ARM-ready-with-multi-arch-container-builds.md [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+restartPolicy: Always
+nodeSelector:
+  kubernetes.io/arch: arm64
+```
+
+The important fields to notice here are:
+* `spec.containers.images: arm64v8/busybox` - notice the arm specification.
+* `nodeSelector.kubernetes.io/arch: arm64` - this tells kubernetes to run the pod on an arm64 node.
+
+The label `kubernetes.io/arch` is supplied to the nodes when they are added in EKS to the node group. This is specified so that if non-arm nodes are added to the cluster, the pod will still get assigned to the matching node based on the architecture (arch for sure). Authenticate to the cluster and run `kubectl apply -f busybox-arm64.yaml`. This will deploy this pod to a node containing the `arm64` label, as desired.
```
Comment (fhatfield-ippon) [inline posts/Modernize-your-pipeline-to-be-ARM-ready-with-multi-arch-container-builds.md:85]: Is "(arch for sure)" supposed to say "(arch, for short)"?

File: posts/Modernize-your-pipeline-to-be-ARM-ready-with-multi-arch-container-builds.md [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+```
+
+Next, create a `Dockerfile` in the same directory as your python program
+
+```Dockerfile
+FROM python
+COPY ./print-platform-info.py /print-platform-info
+CMD ["python3", "/print-platform-info"]
+```
+Now this is typically where you would scamper off to the command line and run `docker build -t my_image .`, but hold off for just a minute! We have to set docker to use a different *build instance*. In the command line inside the project directory where you have your `Dockerfile` run:
```
Comment (fhatfield-ippon) [inline posts/Modernize-your-pipeline-to-be-ARM-ready-with-multi-arch-container-builds.md:106]: Isn't it wild that "hold off" and "hold on" would mean the same thing in this context?

**Comment by fhatfield-ippon** [comment]
I just had some tiny tweaks/suggestions, but otherwise, as usual, your blog is super approachable and informative, and I enjoyed reading it. Nice job!

File: posts/Modernize-your-pipeline-to-be-ARM-ready-with-multi-arch-container-builds.md (hunk 8 above)
Comment (lward27) [reply posts/Modernize-your-pipeline-to-be-ARM-ready-with-multi-arch-container-builds.md:106]: Lol good ole English language..
